### PR TITLE
Create a python app for dependencies instead of venv

### PR DESCRIPTION
**Pull request by mppf**
Follow-up to PR #16663 #16644 and #16560

We run into problems when including a venv in the installed code for
Chapel (e.g. in a Chapel RPM) because the venv encodes the path to the
python interpreter (both as symbolic links and also in pyvenv.cfg). We
tried to work around this in PRs #3571 #16663 and #16680.

This PR instead changes the strategy away from trying to create a
long-lasting venv in the `third-party/chpl-venv`. Instead, it creates a
venv during installation (so we can have `pip` even on systems that don't
include it by default) and then uses that venv to install the
dependencies into `chpldeps` which is a Python application. Now we
install the dependencies with `python3 -m pip --target path/to/chpldeps`.

Additionally we no longer include `getpip` support in the Makefiles as it
should no longer be used and we stop passing `--cache-dir` to the `pip`
commands (so that the cache can be used night-to-night in nightly
testing).

Originally I had hoped to use `zipapp` to store `chpldeps` as a zipfile
but that causes problems with Sphinx (when it looks for theme files). So,
here we just have `chpldeps` as a directory (which Python considers a
kind of application) just as we would before zipping it with `zipapp`.
`zipapp` does not do much other than storing the files into a .zip and so
this should have similar portability.

To keep it convenient, Python programs needing `chpldeps` can be run as
`python3 path/to/chpldeps some/python/program.py` or for certain included
tools, `python3 path/to/chpldeps sphinx-build`. This avoids the need for
`util/config/run-in-venv.bash` in many cases.

Lastly, I wasn't sure what to do with chplspell for this PR since
chplspell is a developer-only tool. For now, it uses the same venv that
we install pip into.

Reviewed by @ben-albrecht and @lydia-duncan - thanks!

- [x] full local testing
- [x] `test_rpm_module.bash`

### CODE REVIEW

File: third-party/chpl-venv/Makefile.include [hunk 1]
```diff
@@ -8,7 +8,6 @@ CHPL_VENV_DIR=$(shell cd $(THIRD_PARTY_DIR)/chpl-venv && pwd)
 CHPL_VENV_TEST_REQUIREMENTS_FILE=$(CHPL_VENV_DIR)/test-requirements.txt
 CHPL_VENV_CHPLDOC_REQUIREMENTS_FILE=$(CHPL_VENV_DIR)/chpldoc-requirements.txt
 CHPL_VENV_C2CHAPEL_REQUIREMENTS_FILE=$(CHPL_VENV_DIR)/c2chapel-requirements.txt
-CHPL_VENV_CHPLSPELL_REQUIREMENTS_FILE=$(CHPL_VENV_DIR)/chplspell-requirements.txt
```
Comment (lydia-duncan) [inline third-party/chpl-venv/Makefile.include:11]: Why have we removed chplspell stuff?

**Review by lydia-duncan — COMMENTED** [review]
What is the behavior:
- on a Mac after a clobber?
- on Linux after a clobber?
- when calling `make chpldoc`/`make test-venv` when nothing has changed in chpl-venv since the last time it was run?
- with a `make docs` call when chpldoc is not built?
- with a `make docs` call when chpldoc *has* been built?

**Comment by mppf** [comment]
>    on a Mac after a clobber?

`make docs` downloads the dependencies. a second `make docs` does not download. both succeed.

>    on Linux after a clobber?

ditto

>    when calling make chpldoc/make test-venv when nothing has changed in chpl-venv since the last time it was run?

nothing is downloaded

>    with a make docs call when chpldoc is not built?

it builds chpldoc. it does not rebuild the chpldeps if they have not changed.

>    with a make docs call when chpldoc has been built?

nothing is downloaded because `make chpldoc` gathers the dependencies and constructs chpldeps.

File: third-party/chpl-venv/chpldeps-main.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+import os
+import re
+import sys
+import importlib.util
+
```
Comment (ben-albrecht) [inline third-party/chpl-venv/chpldeps-main.py:5]: This file could use a short doc string at the top saying what it's all about.

File: util/chplenv/chpl_home_utils.py [hunk 3]
```diff
@@ -23,22 +23,12 @@ def get_chpl_third_party():
     return chpl_third_party
 
 # Get the chpl-venv install directory:
-# $CHPL_HOME/third-party/chpl-venv/install/$bin_subdir/$py_version/chpl-virtualenv
+# $CHPL_HOME/third-party/chpl-venv/install/chpldeps.pyz
```
Comment (ben-albrecht) [inline util/chplenv/chpl_home_utils.py:26]: is `pyz` intentional here?

**Pull request merged**